DGMulti improvements and preparation for Gauss collocation solver

jlchan opened the pull request.

jlchan commented:
@ranocha @sloede this was intended to extend solver capabilities to Gauss collocation, but it has resulted in a fair number of "clean-up" commits already. I'm thinking of breaking this work into 3 PRs

1. "clean-up" and prep work (this PR)
2. Switch to a trait system for DGMulti solvers
3. Add Gauss collocation/generalized SBP solvers using the trait system.

Thoughts or feedback?

sloede commented:
> @ranocha @sloede this was intended to extend solver capabilities to Gauss collocation, but it has resulted in a fair number of "clean-up" commits already. I can get these "clean-up" changes ready for review, then implement Gauss collocation in a new PR, or I can continue implementing things in this one PR. Do you have a preference?

My personal preference is clear: As long as it is a logically complete, non-trivial change set, the smaller the PR the better. I'll have a look at this later today!

ranocha commented:
Same for me: Please clean this up and ping me when you're ready. It's much easier to review smaller PRs and will let us move forward faster.

jlchan commented:
@sloede @ranocha This PR mainly adds a Rayleigh-Taylor simulation for DGMulti + fixes bugs and performance issues associated with it.

1. Added Rayleigh-Taylor for DGMulti (added a new `calc_sources` to fix a bug with `approximation_type=SBP()`)
2. Performance improvements (use sparsity-based flux differencing by default on tensor product elements, parallelize over evaluations of `entropy2cons` and `cons2entropy`)
3. Minor bug fixes (for DGMulti) and improvements in Makie plot recipes (specify default colormap)

The one structural change is that I added a new "dispatchable" constructor for DGMulti. This will allow customizable setups, which I'll use for the Gauss collocation solver.

sloede commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -35,7 +35,8 @@
 
 @inline function hadamard_sum!(du, A, volume_flux, orientation, u,
-                               equations, sparsity_pattern::AbstractSparseMatrix{Bool})
+                               equations, sparsity_pattern::AbstractSparseMatrix{Bool},
+                               skip_index = nothing)
```
```suggestion
                               equations, sparsity_pattern::AbstractSparseMatrix{Bool})
```
`skip_index` seems to be unused

jlchan replied:
It is, but I introduced it to match the function signature for dispatch in https://github.com/trixi-framework/Trixi.jl/blob/a95136704549584b519a6096b3f0bcab1010e348/src/solvers/dgmulti/flux_differencing.jl#L258. 

The type of `sparsity_pattern` toggles dispatch and overrides `skip_index` if `sparsity_pattern != nothing`

sloede commented on src/solvers/dgmulti/types.jl:
```diff
@@ -41,6 +40,21 @@ function DGMulti(; polydeg::Integer,
+end
+
+# dispatchable constructor for DGMulti to allow for specialization
+function DGMulti(element_type::AbstractElemShape,
+                 approximation_type,
+                 volume_integral=VolumeIntegralWeakForm(),
+                 surface_integral=SurfaceIntegralWeakForm(surface_flux);
+                 polydeg::Integer,
+                 surface_flux=flux_central,
+                 kwargs...)
```
If I understand you correctly, you mainly need this for internal use to allow dispatching. Personally, I think it's not very good practice to have default values defined in multiple, non-orthogonal places (if a person in the future decides to change one default value in one place, it might forget about the other places). Thus I suggest to remove the default values from here.
```suggestion
function DGMulti(element_type::AbstractElemShape,
                 approximation_type,
                 volume_integral,
                 surface_integral;
                 polydeg::Integer,
                 surface_flux,
                 kwargs...)
```

jlchan replied:
That sounds like a good idea. Thanks!

sloede commented on src/visualization/recipes_makie.jl:
```diff
@@ -108,7 +108,9 @@ iplot(u, semi; kwargs...) = iplot(wrap_array_native(u, semi), mesh_equations_sol
 @Makie.recipe(TrixiHeatmap, plot_data_series) do scene
-  Makie.Theme(;)
+  Makie.Theme(
+    colormap = :thermal,
```
Why change the default?

jlchan replied:
It looks more like the default `:thermal` for Plots.jl. Couldn't find an exact match :(

ranocha replied:
Should we change our default to `:plasma` for Plots.jl? I guess this should also be available for Makie.jl, since it's one of the three "big" colormaps from PyPlot (viridis, plasma, magma)

ranocha replied:
Here are some "standard" scientific colormaps
![image](https://user-images.githubusercontent.com/12693098/130496339-1d996e99-16a5-40af-a7bf-84988d8c99ad.png)
Inferno is basically the standard of Plots.jl.

jlchan replied:
Hm - let me try `:inferno` again. I think I misread http://makie.juliaplots.org/v0.15.1/documentation/colors/index.html - :inferno is actually under matplotlib.

jlchan replied:
Ah, it works! Let me add :inferno as default to 100% match Plots.jl for now. Later, it might be nice to add a `default_colormap` function to make this configurable.

sloede commented on src/visualization/utilities.jl:
```diff
@@ -1126,7 +1126,7 @@ function mesh_plotting_wireframe(pds::PlotDataSeries{<:UnstructuredPlotData2D};
 
   if set_z_coordinate_zero
-    sol_f = zeros(eltype(first(face_data)), size(face_data)) # plot 2d surface by setting z coordinate to zero
+    sol_f = zeros(eltype(first(x_face)), size(x_face)) # plot 2d surface by setting z coordinate to zero
```
Why this change?

jlchan replied:
I set `face_data` as nothing for Plots.jl. It's used to set the z-coordinate for 3D surface plots, which aren't used for Plots.jl recipes.

sloede commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+                                                               slope=1000)
+  tol = 1e2*eps()
+
+  if x[2] < .5
```
```suggestion
  if x[2] < 0.5
```

sloede commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+vy = map(x-> 0.5 * (1+x), vy) # map to [0, 1] for single mode RTI
+vertex_coordinates = (vx, vy)
+
+# boundary_conditions = (; :entire_boundary => BoundaryConditionDirichlet(initial_condition))
```
```suggestion
```
May be removed?

jlchan replied:
Oops - yes! Thanks.

sloede reviewed (changes requested):
Generally looks good to me - and small enough to do a quick review :+1: I have left a few comments...

ranocha commented on src/visualization/recipes_makie.jl:
```diff
@@ -108,7 +108,9 @@ iplot(u, semi; kwargs...) = iplot(wrap_array_native(u, semi), mesh_equations_sol
-  Makie.Theme(;)
+  Makie.Theme(
+    colormap = :thermal,
+  )
```
A related question that comes to my mind here (since I'm testing it): Is it possible to add the colorbars to the Makie plots as we do for the Plots plots? For example, `GLMakie.plot(sol, plot_mesh=false)` gives me the plot below, which is hard to interpret without colorbars.

![image](https://user-images.githubusercontent.com/12693098/130495161-8c8391c6-74c7-4bd7-a776-6b9f0ffac181.png)

jlchan replied:
Yeah, let me give that a shot. I tried before and it looked odd, but I think I have to do something with Makie's "Aspect" tools.

ranocha replied:
If it's not easy to do right now, please open an issue for it

jlchan replied:
Sounds good.

ranocha commented:
Ha looks like @sloede and I were reviewing this basically at the same time and left quite often basically the same comments :sweat_smile:

jlchan commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+
+num_elements = 16
+cells_per_dimension = (num_elements, 4 * num_elements)
+vertex_coordinates, EToV = StartUpDG.uniform_mesh(dg.basis.elementType, cells_per_dimension...)
+
+vx, vy = vertex_coordinates
+vx = map(x-> 0.25 * 0.5*(1 + x), vx) # map to [0, .25]
+vy = map(x-> 0.5 * (1+x), vy) # map to [0, 1] for single mode RTI
+vertex_coordinates = (vx, vy)
```
> The `StartUpDG.uniform_mesh` is on `[-1, 1]^ndims`, correct?

That's correct

jlchan commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+vy = map(x-> 0.5 * (1+x), vy) # map to [0, 1] for single mode RTI
+vertex_coordinates = (vx, vy)
+
+# boundary_conditions = (; :entire_boundary => BoundaryConditionDirichlet(initial_condition))
```
I can remove it - thanks.

jlchan commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+# boundary_conditions = (; :entire_boundary => BoundaryConditionDirichlet(initial_condition))
+boundary_conditions = (; :entire_boundary => BoundaryConditionWall(boundary_state_slip_wall))
+
+mesh = VertexMappedMesh(vertex_coordinates, EToV, dg; is_periodic=(true,false))
```
Good idea, yes.

jlchan commented on src/solvers/dgmulti/dg.jl:
```diff
@@ -9,7 +9,7 @@
 
 # out <- out + A * x
-mul_by_accum!(A) = @inline (out, x)->matmul!(out, A, x, one(eltype(out)), one(eltype(out)))
+mul_by_accum!(A) = @inline (out, x)->mul!(out, A, x, one(eltype(out)), one(eltype(out)))
```
I was running into issues similar to https://github.com/JuliaLinearAlgebra/Octavian.jl/issues/108 again. I'm trying to reproduce it in an MWE so I can reopen the issue, but I haven't gotten around to it.

How about adding a comment referencing https://github.com/JuliaLinearAlgebra/Octavian.jl/issues/108 here?

ranocha replied:
Sounds good :+1:

jlchan commented on src/solvers/dgmulti/dg.jl:
```diff
@@ -118,6 +119,27 @@ function estimate_dt(mesh::AbstractMeshData, dg::DGMulti)
+    end
+    dt_min = min(dt_min, h_e / sum(max_speeds))
+  end
+  domain_size = StartUpDG.compute_domain_size(rd, md)^(1/NDIMS)
```
That is a very good point...I can remove it. The scaling factor is just `2` in any case for a hypercube on `[-1,1]^NDIMS`.

jlchan commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -35,7 +35,8 @@
 end
 
 @inline function hadamard_sum!(du, A, volume_flux, orientation, u,
-                               equations, sparsity_pattern::AbstractSparseMatrix{Bool})
+                               equations, sparsity_pattern::AbstractSparseMatrix{Bool},
+                               skip_index = nothing)
   n = size(sparsity_pattern, 2)
   rows = rowvals(sparsity_pattern)
```
Sounds good - I call it with `skip_index::Function` in https://github.com/trixi-framework/Trixi.jl/blob/a95136704549584b519a6096b3f0bcab1010e348/src/solvers/dgmulti/flux_differencing.jl#L258, and I'll add that to the signature.

ranocha replied:
So you're adding this as a preparation step here since you will implement another method including `skip_index` later?

jlchan replied:
Ah, no - I actually forgot a use case. This is meant to catch three different cases

1. `skip_index::Nothing`
2. `skip_index::Function`
3. `skip_index` not specified

jlchan replied:
Basically if `sparsity_pattern::AbstractSparseMatrix` is specified, then `skip_index` is ignored.

jlchan replied:
I've added the following fix:
```julia
# If `skip_index` isn't specified, set it to `nothing` and dispatch based on `sparsity_pattern`
function hadamard_sum!(du, A, volume_flux, orientation, u, equations, sparsity_pattern)
  return hadamard_sum!(du, A, volume_flux, orientation, u, equations, sparsity_pattern, nothing)
end

function hadamard_sum!(du, A, volume_flux, orientation, u,
                                          equations, sparsity_pattern::AbstractSparseMatrix{Bool},
                                          skip_index::Union{Function, Nothing})
```
Please let me know if you have any feedback.

jlchan commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -202,11 +191,15 @@ function entropy_projection!(cache, u, mesh::VertexMappedMesh, equations, dg::DG
   # TODO: simplices. Address hard-coding of `entropy2cons!`
   apply_to_each_field(mul_by!(Vq), u_values, u)
-  entropy_var_values .= cons2entropy.(u_values, equations)
+  @threaded for i in Base.OneTo(length(u_values))
+    entropy_var_values[i] = cons2entropy(u_values[i], equations)
```
I didn't know about `eachindex` - that is *very* convenient. Thanks!

jlchan commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+using Trixi, OrdinaryDiffEq
```
Correct

jlchan commented on examples/dgmulti_2d/elixir_euler_rayleigh_taylor_instability.jl:
```diff
@@ -0,0 +1,98 @@
+using Trixi, OrdinaryDiffEq
```
Good idea, and sure.

jlchan replied:
Adding KHI instability too (apparently it wasn't there already)

jlchan commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -202,11 +191,15 @@ function entropy_projection!(cache, u, mesh::VertexMappedMesh, equations, dg::DG
   # TODO: simplices. Address hard-coding of `entropy2cons!`
   apply_to_each_field(mul_by!(Vq), u_values, u)
-  entropy_var_values .= cons2entropy.(u_values, equations)
+  @threaded for i in eachindex(u_values)
```
Odd! `eachindex` causes the simulation to crash, but it runs with `Base.OneTo(length(...))`.

ranocha replied:
Which types do `entropy_var_values` and `u_values` have?

jlchan replied:
@ranocha adding a comment on this. Seems like an issue with Polyester.

jlchan replied:
They're both StructArray types I think, so `eachindex` returns iterators over `CartesianIndex`

jlchan replied:
Opened an issue https://github.com/JuliaSIMD/Polyester.jl/issues/37

Mind if I just add a comment linking to the issue for now?

sloede commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -193,20 +187,27 @@ function create_cache(mesh::VertexMappedMesh, equations, dg::DGMultiFluxDiff{<:P
   apply_to_each_field(mul_by!(Vq), u_values, u)
-  entropy_var_values .= cons2entropy.(u_values, equations)
+
+  # TODO: DGMulti. @threaded crashes when using `eachindex(u_values)`.
+  # See https://github.com/JuliaSIMD/Polyester.jl/issues/37 for more details.
+  for i in Base.OneTo(length(u_values))
```
```suggestion
  # TODO: DGMulti. @threaded crashes when using `eachindex(u_values)`.
  # See https://github.com/JuliaSIMD/Polyester.jl/issues/37 for more details.
  @threaded for i in Base.OneTo(length(u_values))
```
But you're not using `eachindex` anymore, so adding `@threaded` here should be fine, shouldn't it?

jlchan replied:
Oops - yes! I accidentally committed my experimentation...

sloede commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -268,6 +270,20 @@ function calc_volume_integral!(du, u, volume_integral,
 
+# Specialize since `u_values` isn't computed for DGMultiFluxDiff{<:SBP} solvers.
+function calc_sources!(du, u, t, source_terms,
+                       mesh::AbstractMeshData, equations, dg::DGMultiFluxDiff{<:SBP}, cache)
```
```suggestion
                       mesh, equations, dg::DGMultiFluxDiff{<:SBP}, cache)
```
Same as above

sloede reviewed (changes requested):
Just a few small remarks left, but overall, this LGTM (looks good to me/merge)

jlchan commented on src/solvers/dgmulti/dg.jl:
```diff
@@ -273,20 +297,23 @@ function calc_single_boundary_flux!(cache, t, boundary_condition, boundary_key,
+              mesh::AbstractMeshData, equations, dg::DGMulti, cache) = nothing
 calc_sources!(du, u, t, source_terms::Nothing,
-              mesh::VertexMappedMesh, equations, dg::DGMulti, cache) = nothing
+              mesh::AbstractMeshData, equations, dg::DGMultiFluxDiff{<:SBP}, cache) = nothing
```
I think it might be OK - trying it out.

sloede replied:
If it gives you any headache, just leave it as it is...

jlchan replied:
Seems like it works fine.

sloede reviewed (approved):
LGTM once tests pass & coverage is ok

ranocha reviewed (approved):
Thanks!
